src/mcpwm.rs

```diff
@@ -0,0 +1,527 @@
+//! }
+//! ```
+//!
+//! See the `examples/` folder of this repository for more.
+
+use core::borrow::Borrow;
+
+use crate::gpio::OutputPin;
+use crate::units::{FromValueType, Hertz};
+use esp_idf_sys::{esp, mcpwm_config_t, EspError};
+use esp_idf_sys::{
+    mcpwm_counter_type_t,
+    mcpwm_counter_type_t_MCPWM_DOWN_COUNTER,
+    //mcpwm_counter_type_t_MCPWM_FREEZE_COUNTER,
+    mcpwm_counter_type_t_MCPWM_UP_COUNTER,
+    mcpwm_counter_type_t_MCPWM_UP_DOWN_COUNTER,
+};
+use esp_idf_sys::{
+    mcpwm_duty_type_t, mcpwm_duty_type_t_MCPWM_DUTY_MODE_0, mcpwm_duty_type_t_MCPWM_DUTY_MODE_1,
+};
+use esp_idf_sys::{
+    mcpwm_io_signals_t, mcpwm_io_signals_t_MCPWM0A, mcpwm_io_signals_t_MCPWM0B,
+    mcpwm_io_signals_t_MCPWM1A, mcpwm_io_signals_t_MCPWM1B, mcpwm_io_signals_t_MCPWM2A,
+    mcpwm_io_signals_t_MCPWM2B,
+};
+use esp_idf_sys::{
+    mcpwm_timer_t, mcpwm_timer_t_MCPWM_TIMER_0, mcpwm_timer_t_MCPWM_TIMER_1,
+    mcpwm_timer_t_MCPWM_TIMER_2,
+};
+use esp_idf_sys::{mcpwm_unit_t, mcpwm_unit_t_MCPWM_UNIT_0, mcpwm_unit_t_MCPWM_UNIT_1};
+
+/// The Motor Control Pulse Width Modulator peripheral
+pub struct Peripheral<U: Unit> {
+    pub mcpwm: MCPWM<U>, // TODO: Is there a better way to name or structure this?
+    pub operator0: OPERATOR0<U>,
+    pub operator1: OPERATOR1<U>,
+    pub operator2: OPERATOR2<U>,
+}
+
+impl<U: Unit> Peripheral<U> {
+    /// # Safety
+    ///
+    /// It is safe to instantiate this exactly one time per `Unit`.
+    pub unsafe fn new() -> Self {
+        Self {
+            mcpwm: MCPWM::new(),
+            operator0: OPERATOR0::new(),
+            operator1: OPERATOR1::new(),
+            operator2: OPERATOR2::new(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum DutyMode {
+    ActiveHigh,
+    ActiveLow,
+}
+
+impl From<DutyMode> for mcpwm_duty_type_t {
+    fn from(val: DutyMode) -> Self {
+        match val {
+            DutyMode::ActiveHigh => mcpwm_duty_type_t_MCPWM_DUTY_MODE_0,
+            DutyMode::ActiveLow => mcpwm_duty_type_t_MCPWM_DUTY_MODE_1,
+        }
+    }
+}
+
+// TODO: For UpDown, frequency is half of MCPWM frequency set
+#[derive(Clone, Copy, Debug)]
+pub enum CounterMode {
+    //Frozen,
+    Up,
+    Down,
+    UpDown,
+}
+
+impl From<CounterMode> for mcpwm_counter_type_t {
+    fn from(val: CounterMode) -> Self {
+        match val {
+            // TODO: This seems to be new to IDF 4.4?
+            //CounterMode::Frozen => mcpwm_counter_type_t_MCPWM_FREEZE_COUNTER,
+            CounterMode::Up => mcpwm_counter_type_t_MCPWM_UP_COUNTER,
+            CounterMode::Down => mcpwm_counter_type_t_MCPWM_DOWN_COUNTER,
+            CounterMode::UpDown => mcpwm_counter_type_t_MCPWM_UP_DOWN_COUNTER,
+        }
+    }
+}
+
+pub struct OperatorConfig {
+    frequency: Hertz,
+    duty_a: Duty,
+    duty_b: Duty,
+
+    #[cfg(idf_newer_than_or_equal_v4_4_0)]
+    lowest_frequency: Hertz,
+    duty_mode: DutyMode,
+    counter_mode: CounterMode,
+    //deadtime: DeadtimeConfig,
```

Comment:
Do you plan on implementing support for dead time? Seems quite useful for some applications.

Reply:
To be clear, in my mind this still quite an early draft :). Just wanted to post some rough code to get some early indication on whether I am heading in the right direction.

As for deadtime, syncronization, and fault detect/capture, I think I would like to implement atleast one of them in this PR. Feels like it would be nice to have atleast something to show of what makes MCPWM unique from LEDC. 

Longer term, I guess everything, right? :)

Reply:
Dead time support is now added.

I think the other ones will have to wait for another PR(s). Some of the features such as synchronization I am a bit unsure how to best expose. Should the operator somehow keep a reference to the sync source in order to guarantee that the sync source does not disappear while in use. We can not take ownership of the sync source since there may be multiple operators using the same sync source. Should a synchronized operator have its own wrapper with a method for adjusting the sync offset/phase etc in addition to the regular `Operator` methods?

There is also a similar situation for "fault" and "capture" with regards to having a reference to the source.

Anyways, lots to think about for future PRs :)